fullstack-cert/typescript-projects/fortune-app/src/script.ts

```diff
@@ -0,0 +1,265 @@
+  elements.forEach((el) => el.classList.remove("hidden"));
+
+/** Utility: pick a random item from a non-empty array */
+const getRandomItem = <T>(items: T[]): T => {
+  const index = Math.floor(Math.random() * items.length);
+  return items[index];
+};
+
+/** Local fallback image for cards that fail to load a remote image */
+const LOCAL_DEFAULT_IMG = defaultImg; // Place default.svg in your public or assets folder
+
+/**
+ * Render a single card block with a loader, image, and error fallback.
+ * - Uses onload to hide the loader once the image is ready
+ * - Uses onerror to retry once with a local default image
+ */
+const renderCard = (
+  drawingType: string,
+  isReversed: boolean,
+  shortName: string,
+  img: string,
+): string =>
+  `
+  <div>
+    <h2>${drawingType}</h2>
+    <div class="card_container ${isReversed ? "reversed-card" : ""}" data-id="${shortName}">
+      <div class="img-loader"></div>
+      <img 
+        src="${img ? `${CDN_URL}/${img}` : LOCAL_DEFAULT_IMG}"
+        class="card-img hidden"
+        onload="this.classList.remove('hidden');this.previousElementSibling.style.display='none';"
+        onerror="
+          if (!this.dataset.failed) {
+            this.dataset.failed = '1';
+            this.src='${LOCAL_DEFAULT_IMG}';
+          } else {
+            this.classList.remove('hidden');
+            this.previousElementSibling.style.display='none';
+          }
+        "
+      />
+    </div>
+  </div>
+`;
+
+/** Types of card positions when drawing multiple cards */
+enum DrawingType {
+  Past = "past",
+  Present = "present",
+  Future = "future",
+}
+
+/**
+ * Game
+ * - Holds card data, DOM references, event bindings, and UI transitions
+ */
+class Game {
+  cards: Card[] = [];
+  /** Cache all relevant DOM elements up front */
+  private elements: {
+    singleCardBtn: HTMLElement;
+    singleCard: HTMLElement;
+    multipleCardsBtn: HTMLElement;
+    multipleCard: HTMLElement;
+    title: HTMLElement;
+    newReadingBtn: HTMLElement;
+    fortuneContainer: HTMLElement;
+    fortuneDescription: HTMLElement;
+    headerTitle: HTMLElement;
+    subTitle: HTMLElement;
+    cardTitle: HTMLElement;
+    description: HTMLElement;
+    text: HTMLElement;
+  };
+
+  constructor() {
+    // 1) Cache DOM elements
+    this.elements = {
+      singleCardBtn: getElement("#btn-single-card")!,
+      singleCard: getElement(".single_card")!,
+      multipleCardsBtn: getElement("#btn-multiple-cards")!,
+      multipleCard: getElement(".multiple_card")!,
+      title: getElement(".title")!,
+      newReadingBtn: getElement(".btn_reveal")!,
+      fortuneContainer: getElement(".fortune_container")!,
+      fortuneDescription: getElement(".fortune_description")!,
+      headerTitle: getElement(".header_title")!,
+      subTitle: getElement(".sub_title")!,
+      cardTitle: getElement(".desc_title")!,
+      description: getElement(".description")!,
+      text: getElement(".text")!,
```

Comment:
```suggestion
      singleCardBtn: getElement("#btn-single-card")!,
      singleCard: getElement(".single_card")!,
      multipleCardsBtn: getElement("#btn-multiple-cards")!,
      multipleCard: getElement(".multiple_card")!,
      title: getElement(".title")!,
      newReadingBtn: getElement(".btn_reveal")!,
      fortuneContainer: getElement(".fortune_container")!,
      fortuneDescription: getElement(".fortune_description")!,
      headerTitle: getElement(".header_title")!,
      subTitle: getElement(".sub_title")!,
      cardTitle: getElement(".desc_title")!,
      description: getElement(".description")!,
      text: getElement(".text")!,
```

Rather than non-null assertions, we should implement some sort of fallback or error handling.